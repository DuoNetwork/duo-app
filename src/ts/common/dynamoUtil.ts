import { DynamoUtil } from '@finbook/duo-market-data';

let config = {};
try {
	config = require(`../keys/aws.ui.${__KOVAN__ ? 'dev' : 'live'}.json`);
} catch (error) {
	console.log(error);
}

const dynamoUtil = new DynamoUtil(config, !__KOVAN__, '');

export default dynamoUtil;
