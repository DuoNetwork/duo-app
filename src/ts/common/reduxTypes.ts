import { IAssets, IPriceData, ITimeSeriesData } from './types';
export interface IState {
	ui: IUIState;
}

export interface IUIState {
	eth: ITimeSeriesData[];
	classA: ITimeSeriesData[];
	classB: ITimeSeriesData[];
	reset: ITimeSeriesData[];
	history: string[];
	message: {
		type: string;
		content: string;
		visible: boolean;
	};
	setting: {
		couponRate: number;
		upwardResetLimit: number;
		downwardResetLimit: number;
		periodicResetLimit: number;
	};
	form: {
		type: string;
		visible: boolean;
	};
	mv: ITimeSeriesData[];
	assets: IAssets;
	resetPrice: number;
	beta: number;
	day: number;
	upward: number;
	downward: number;
	periodic: number;
	price: IPriceData;
}

export type Action =
	| IBaseAction
	| IBooleanAction
	| IStringAction
	| IObjectAction
	| INumberAction
	| ITradeAction;

export type Dispatch = (action: Action | ThunkAction | PromiseAction) => any;

export interface IBaseAction {
	type: string;
}

export interface IBooleanAction {
	type: string;
	value: boolean;
}

export interface IStringAction {
	type: string;
	value: string;
}

export interface IObjectAction {
	type: string;
	value: object;
}

export interface INumberAction {
	type: string;
	value: number;
}

export interface ITradeAction {
	type: string;
	history: string;
	assets: IAssets;
}

export type ThunkAction = (dispatch: Dispatch) => any;
export type PromiseAction = Promise<Action>;
