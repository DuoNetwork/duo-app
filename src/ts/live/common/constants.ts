export * from '../../../../../duo-admin/src/constants';

export const AC_CTD_STATES = 'custodianStates';
export const AC_CTD_PRICES = 'custodianPrices';
export const AC_ACCOUNT = 'account';
export const AC_BALANCES = 'balances';
export const AC_ADDRESSES = 'addresses';
export const AC_DNM_STATUS = 'dynamoStatus';
export const AC_DMN_HOURLY = 'dynamoHourly';
export const AC_DMN_MINUTELY = 'dynamoMinutely';
export const AC_DMN_PRICES = 'dynamoPrices';
export const AC_REFRESH = 'refresh';
