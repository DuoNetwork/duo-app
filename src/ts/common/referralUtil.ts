import { AttributeMap } from 'aws-sdk/clients/dynamodb';
import moment from 'moment';
import dynamoUtil from './dynamoUtil';
import * as StakingCST from './stakingCST';
import { IAddressInfo, IReferral, IStakingChild, IStakingNode } from './types';
class ReferralUtil {
	public getUTCNowTimestamp() {
		return moment().valueOf();
	}

	public async checkExist(account: string) {
		const data = await dynamoUtil.scanData({ TableName: StakingCST.REFERRALTABLE });
		const addressList: string[] = [];
		if (data.Items)
			data.Items.map(item => {
				addressList.push((item as any).address.S);
			});
		if (addressList.includes(account)) {
			const index = addressList.indexOf(account);
			return (data.Items as any)[index].referralCode.S;
		}
		return '';
	}

	public async insertReferralEntry(item: IReferral) {
		const data: AttributeMap = {
			address: { S: item.address },
			referralCode: { S: item.referralCode },
			signHash: { S: item.signHash },
			updatedAt: { S: this.getUTCNowTimestamp() + '' }
		};
		const params = {
			TableName: StakingCST.REFERRALTABLE,
			Item: data
		};

		await dynamoUtil.insertData(params);
	}

	public async getAddressInfo(account: string) {
		const params = {
			TableName: StakingCST.STAKINGTABLE,
			KeyConditionExpression: 'address = :address',
			ExpressionAttributeValues: {
				[':address']: { S: account }
			}
		};
		const data = await dynamoUtil.queryData(params);
		if (data.Count) {
			const rawRA = (data.Items as any)[0].referralAward;
			const rawS0 = (data.Items as any)[0].staking0Award;
			const rawS60 = (data.Items as any)[0].staking60Award;
			const addressInfo: IAddressInfo = {};
			if (rawRA)
				if ((rawRA as any).S !== 'no referee') {
					const rawChildList = rawRA.S.split(';');
					const childList: IStakingChild[] = [];
					rawChildList.forEach((child: string) => {
						const rawData = child.split(',');
						const childData = {
							address: rawData[0],
							daily: Number(rawData[1]),
							accumulated: Number(rawData[2])
						};
						childList.push(childData);
					});
					addressInfo.children = childList;
				}
			if (rawS0) {
				const rawNodeList = rawS0.S.split(';');
				const nodeList: IStakingNode[] = [];
				rawNodeList.forEach((node: string) => {
					const rawData = node.split(',');
					const nodeData = {
						name: rawData[0],
						daily: Number(rawData[1]),
						accumulated: Number(rawData[2])
					};
					nodeList.push(nodeData);
				});
				addressInfo.staking0 = nodeList;
			}
			if (rawS60) {
				const rawNodeList = rawS60.S.split(';');
				const nodeList: IStakingNode[] = [];
				rawNodeList.forEach((node: string) => {
					const rawData = node.split(',');
					const nodeData = {
						name: rawData[0],
						daily: Number(rawData[1]),
						accumulated: Number(rawData[2])
					};
					nodeList.push(nodeData);
				});
				addressInfo.staking60 = nodeList;
			}
			return addressInfo;
		} else return {};
	}
}

const referralUtil = new ReferralUtil();
export default referralUtil;
