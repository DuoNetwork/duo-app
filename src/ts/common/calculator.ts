import moment from 'moment';
import * as CST from './constants';
import { IUIState } from './reduxTypes';
import { IAssets, IRawData, ITimeSeriesData } from './types';

class Calculator {
	public getAllTimeSeriesFromEth(
		rawData: IRawData[],
		alpha: number = 1,
		coupon: number = 0.0002,
		upperLimit: number = 2,
		lowerLimit: number = 0.25,
		periodLimit: number = 1.02
	) {
		const eth: ITimeSeriesData[] = [],
			classA: ITimeSeriesData[] = [],
			classB: ITimeSeriesData[] = [],
			resetPrice: ITimeSeriesData[] = [],
			beta: ITimeSeriesData[] = [],
			upward: ITimeSeriesData[] = [],
			downward: ITimeSeriesData[] = [],
			periodic: ITimeSeriesData[] = [];
		let count = -1;
		let lastResetDatetime = moment(rawData[0].Date, 'YYYY-MM-DD').valueOf();
		let lastResetPrice = rawData[0].Price;
		let currentBeta = 1;
		rawData.forEach(d => {
			count++;
			const datetime = moment(d.Date, 'YYYY-MM-DD').valueOf();
			const ethPx = d.Price;
			eth.push({
				datetime: datetime,
				value: ethPx
			});

			let [navA, navB] = this.calculateNav(
				ethPx,
				lastResetPrice,
				alpha,
				currentBeta,
				count,
				coupon
			);

			if (navB >= upperLimit || navB <= lowerLimit || navA >= periodLimit) {
				lastResetPrice = ethPx;
				lastResetDatetime = datetime;
				count = 0;
				if (navB >= upperLimit)
					upward.push({
						datetime: datetime,
						value: 1
					});
				else if (navB <= lowerLimit)
					downward.push({
						datetime: datetime,
						value: 1
					});
				else
					periodic.push({
						datetime: datetime,
						value: 1
					});

				// class B does not reset in periodic reset
				if (navA < periodLimit) {
					classB.push({
						datetime: datetime,
						value: navB
					});
					navB = 1;
					currentBeta = 1;
				} else
					currentBeta = this.updateBeta(currentBeta, ethPx, lastResetPrice, navA, alpha);

				classA.push({
					datetime: datetime,
					value: navA
				});

				navA = 1;
			}

			classA.push({
				datetime: datetime,
				value: navA
			});

			classB.push({
				datetime: datetime,
				value: navB
			});

			resetPrice.push({
				datetime: lastResetDatetime,
				value: lastResetPrice
			});

			beta.push({
				datetime: datetime,
				value: currentBeta
			});
		});
		return [eth, classA, classB, resetPrice, beta, upward, downward, periodic];
	}

	public calculateNav(
		price: number,
		resetPrice: number,
		alpha: number,
		beta: number,
		periodCount: number,
		coupon: number
	) {
		let navParent = price / resetPrice / beta * (1 + alpha);

		let navA = 1 + periodCount * coupon;
		let navAAdj = navA * alpha;
		if (navParent <= navAAdj) return [navParent / alpha, 0];
		else return [navA, navParent - navAAdj];
	}

	public updateBeta(
		prevBeta: number,
		price: number,
		resetPrice: number,
		navA: number,
		alpha: number
	) {
		return (
			(1 + alpha) * price / ((1 + alpha) * price - resetPrice * alpha * prevBeta * (navA - 1))
		);
	}

	public assetUpwardReset(assets: IAssets, eth: number, navA: number, navB: number): IAssets {
		return {
			ETH: assets.ETH + ((navA - 1) * assets.ClassA + (navB - 1) * assets.ClassB) / eth,
			ClassA: assets.ClassA,
			ClassB: assets.ClassB
		};
	}

	public assetDownwardReset(assets: IAssets, eth: number, navA: number, navB: number): IAssets {
		return {
			ETH: assets.ETH + (navA - navB) * assets.ClassA / eth,
			ClassA: assets.ClassA * navB,
			ClassB: assets.ClassB * navB
		};
	}

	public assetPeriodicReset(assets: IAssets, eth: number, navA: number): IAssets {
		return {
			ETH: assets.ETH + (navA - 1) * assets.ClassA / eth,
			ClassA: assets.ClassA,
			ClassB: assets.ClassB
		};
	}

	public calculateNextDayState(state: IUIState): object {
		const { eth, classA, classB, assets, day, upward, downward, periodic, mv } = state;
		const newDayCount = day + 1;
		if (newDayCount >= eth.length) return {};
		else {
			const newDatetime = eth[newDayCount].datetime;
			const newEthPx = eth[newDayCount].value;
			const upwardSoFar = upward.filter(d => d.datetime <= newDatetime);
			const downwardSoFar = downward.filter(d => d.datetime <= newDatetime);
			const periodicSoFar = periodic.filter(d => d.datetime <= newDatetime);
			const isUpward =
				upwardSoFar.length && upwardSoFar[upwardSoFar.length - 1].datetime === newDatetime;
			const isDownward =
				downwardSoFar.length &&
				downwardSoFar[downwardSoFar.length - 1].datetime === newDatetime;
			const isPeriodic =
				periodicSoFar.length &&
				periodicSoFar[periodicSoFar.length - 1].datetime === newDatetime;
			const upwardCount = upwardSoFar.length - (isUpward ? 1 : 0);
			const downwardCount = downwardSoFar.length - (isDownward ? 1 : 0);
			const periodicCount = periodicSoFar.length - (isPeriodic ? 1 : 0);
			const newNavA = classA[newDayCount + upwardCount + downwardCount + periodicCount].value;
			const newNavB = classB[newDayCount + upwardCount + downwardCount].value;

			let newAssets: IAssets;
			let msg = '';
			if (isUpward) {
				newAssets = calculator.assetUpwardReset(assets, newEthPx, newNavA, newNavB);
				msg = "<div style='color: rgba(255,255,255, .8)'>Reset (upward) triggered.</div>";
			} else if (isDownward) {
				newAssets = calculator.assetDownwardReset(assets, newEthPx, newNavA, newNavB);
				msg = "<div style='color: rgba(255,255,255, .8)'>Reset (downward) triggered.</div>";
			} else if (isPeriodic) {
				newAssets = calculator.assetPeriodicReset(assets, newEthPx, newNavA);
				msg = "<div style='color: rgba(255,255,255, .8)'>Reset (periodic) triggered.</div>";
			} else newAssets = assets;

			return Object.assign(
				{
					[CST.AC_MV]: [
						...mv,
						{
							datetime: newDatetime,
							value:
								assets.ETH * newEthPx +
								assets.ClassA * newNavA +
								assets.ClassB * newNavB
						}
					],
					[CST.AC_ASSETS]: newAssets,
					[CST.AC_DAY]: newDayCount,
					[CST.AC_PRICE]: {
						Date: newDatetime,
						ETH: newEthPx,
						ClassA:
							classA[
								newDayCount +
									upwardSoFar.length +
									downwardSoFar.length +
									periodicSoFar.length
							].value,
						ClassB:
							classB[newDayCount + upwardSoFar.length + downwardSoFar.length].value
					}
				},
				msg
					? {
							[CST.AC_MESSAGE]: {
								type: "<div style='color: rgba(0,186,255,0.7)'>INFORMATION</div>",
								content: msg,
								visible: true
							}
					}
					: {}
			);
		}
	}

	public calculateForwardState(state: IUIState): object {
		const { eth, classA, classB, assets, day, upward, downward, periodic, mv } = state;
		const currentDatetime = eth[day].datetime;
		const endDatetime = eth[eth.length - 1].datetime;
		const upwardSoFar = upward.filter(d => d.datetime <= currentDatetime);
		const downwardSoFar = downward.filter(d => d.datetime <= currentDatetime);
		const periodicSoFar = periodic.filter(d => d.datetime <= currentDatetime);
		const nextUpward = upward[upwardSoFar.length]
			? upward[upwardSoFar.length].datetime
			: endDatetime;
		const nextDownward = downward[downwardSoFar.length]
			? downward[downwardSoFar.length].datetime
			: endDatetime;
		const nextPeriodic = periodic[periodicSoFar.length]
			? periodic[periodicSoFar.length].datetime
			: endDatetime;
		const nextResetDate = Math.min(nextUpward, nextDownward, nextPeriodic);
		let curDate = day + 1;
		let newMV = [...mv];
		let newDatetime = eth[curDate].datetime;
		if (newDatetime === nextResetDate) return this.calculateNextDayState(state);

		let newEthPx, newNavA, newNavB;
		while (curDate < eth.length && eth[curDate].datetime < nextResetDate) {
			newDatetime = eth[curDate].datetime;
			newEthPx = eth[curDate].value;
			newNavA =
				classA[curDate + upwardSoFar.length + downwardSoFar.length + periodicSoFar.length]
					.value;
			newNavB = classB[curDate + upwardSoFar.length + downwardSoFar.length].value;
			newMV.push({
				datetime: newDatetime,
				value: assets.ETH * newEthPx + assets.ClassA * newNavA + assets.ClassB * newNavB
			});
			curDate++;
		}
		curDate--;
		return {
			[CST.AC_MV]: newMV,
			[CST.AC_DAY]: curDate,
			[CST.AC_PRICE]: {
				Date: newDatetime,
				ETH: newEthPx,
				ClassA:
					classA[
						curDate + upwardSoFar.length + downwardSoFar.length + periodicSoFar.length
					].value,
				ClassB: classB[curDate + upwardSoFar.length + downwardSoFar.length].value
			}
		};
	}
}

const calculator = new Calculator();
export default calculator;
