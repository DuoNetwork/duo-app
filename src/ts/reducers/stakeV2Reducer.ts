import { Constants as WrapperConstants } from '@finbook/duo-contract-wrapper';
import { AnyAction } from 'redux';
import * as CST from 'ts/common/constants';
import { IStakeV2State } from 'ts/common/types';

export const initialState: IStakeV2State = {
	duo: 0,
	states: {
		stakingEnabled: false,
		lockMinTimeInSecond: 0,
		minStakeAmt: 0,
		maxStakePerOracle: 0,
		totalRewardsToDistribute: 0
	},
	allowance: 0,
	addresses: {
		operator: WrapperConstants.DUMMY_ADDR,
		priceFeedList: []
	},
	subscription: 0,
	userStake: {},
	oracleStake: {},
	userAward: 0,
	contractDUO: 0,
	stagingAdd: {}
};

export function stakeV2Reducer(
	state: IStakeV2State = initialState,
	action: AnyAction
): IStakeV2State {
	switch (action.type) {
		case CST.AC_STK_STATES:
			return Object.assign({}, state, {
				states: action.value
			});
		case CST.AC_STK_ADDRESSES:
			return Object.assign({}, state, {
				addresses: action.value
			});
		case CST.AC_STK_BALANCE:
			return Object.assign({}, state, {
				duo: action.value
			});
		case CST.AC_STK_ALLOWANCE:
			return Object.assign({}, state, {
				allowance: action.value
			});
		case CST.AC_STK_SUB:
			return Object.assign({}, state, {
				subscription: action.id
			});
		case CST.AC_STK_USERSTAKE:
			return Object.assign({}, state, {
				userStake: action.value
			});
		case CST.AC_STK_ORACLESTAKE:
			return Object.assign({}, state, {
				oracleStake: action.value
			});
		case CST.AC_STK_AWARD:
			return Object.assign({}, state, {
				userAward: action.value
			});
		case CST.AC_STK_CONTRACTDUO:
			return Object.assign({}, state, {
				contractDUO: action.value
			});
		case CST.AC_STK_STAGINGADD:
			return Object.assign({}, state, {
				stagingAdd: Object.assign({}, state.stagingAdd, { [action.index]: action.value })
			});
		default:
			return state;
	}
}
