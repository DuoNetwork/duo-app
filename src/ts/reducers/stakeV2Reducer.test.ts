// fix for @ledgerhq/hw-transport-u2f 4.28.0
import '@babel/polyfill';
import { Constants as WrapperConstants } from '@finbook/duo-contract-wrapper';
import * as CST from 'ts/common/constants';
import { initialState, stakeV2Reducer } from './stakeV2Reducer';

describe('stake reducer', () => {
	let state = initialState;

	test('default', () => {
		state = stakeV2Reducer(state, { type: 'any' });
		expect(state).toMatchSnapshot();
	});

	test('stakeStates', () => {
		state = stakeV2Reducer(state, {
			type: CST.AC_STK_STATES,
			value: {
				canStake: false,
				canUnstake: false,
				lockMinTimeInSecond: 300,
				minStakeAmt: 500,
				maxStakePerOracle: 100000,
				totalAwardsToDistribute: 0,
				stakingEnabled: false,
				totalRewardsToDistribute: 0,
			}
		});
		expect(state).toMatchSnapshot();
	});
	test('stakeAddress', () => {
		state = stakeV2Reducer(state, {
			type: CST.AC_STK_ADDRESSES,
			value: {
				operator: WrapperConstants.DUMMY_ADDR,
				priceFeedList: []
			}
		});
		expect(state).toMatchSnapshot();
	});
	test('stakeBalance', () => {
		state = stakeV2Reducer(state, {
			type: CST.AC_STK_BALANCE,
			value: 1000
		});
		expect(state).toMatchSnapshot();
	});
	test('stakeAllowance', () => {
		state = stakeV2Reducer(state, {
			type: CST.AC_STK_ALLOWANCE,
			value: 1000
		});
		expect(state).toMatchSnapshot();
	});
	test('stakeSub On', () => {
		state = stakeV2Reducer(state, {
			type: CST.AC_STK_SUB,
			value: 123
		});
		expect(state).toMatchSnapshot();
	});
	test('stakeSub Off', () => {
		state = stakeV2Reducer(state, {
			type: CST.AC_STK_SUB,
			value: 0
		});
		expect(state).toMatchSnapshot();
	});
	test('userStake', () => {
		state = stakeV2Reducer(state, {
			type: CST.AC_STK_USERSTAKE,
			value: {}
		});
		expect(state).toMatchSnapshot();
	});
	test('oracleStake', () => {
		state = stakeV2Reducer(state, {
			type: CST.AC_STK_ORACLESTAKE,
			value: {}
		});
		expect(state).toMatchSnapshot();
	});
	test('userAward', () => {
		state = stakeV2Reducer(state, {
			type: CST.AC_STK_AWARD,
			value: 0
		});
		expect(state).toMatchSnapshot();
	});
	test('contractDUO', () => {
		state = stakeV2Reducer(state, {
			type: CST.AC_STK_CONTRACTDUO,
			value: 0
		});
		expect(state).toMatchSnapshot();
	});
});
