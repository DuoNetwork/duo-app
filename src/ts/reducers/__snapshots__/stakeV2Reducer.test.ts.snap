// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`stake reducer contractDUO 1`] = `
Object {
  "addresses": Object {
    "operator": "0x0",
    "priceFeedList": Array [],
  },
  "allowance": 1000,
  "contractDUO": 0,
  "duo": 1000,
  "oracleStake": Object {},
  "stagingAdd": Object {},
  "states": Object {
    "canStake": false,
    "canUnstake": false,
    "lockMinTimeInSecond": 300,
    "maxStakePerOracle": 100000,
    "minStakeAmt": 500,
    "stakingEnabled": false,
    "totalAwardsToDistribute": 0,
    "totalRewardsToDistribute": 0,
  },
  "subscription": undefined,
  "userAward": 0,
  "userStake": Object {},
}
`;

exports[`stake reducer default 1`] = `
Object {
  "addresses": Object {
    "operator": "0x0",
    "priceFeedList": Array [],
  },
  "allowance": 0,
  "contractDUO": 0,
  "duo": 0,
  "oracleStake": Object {},
  "stagingAdd": Object {},
  "states": Object {
    "lockMinTimeInSecond": 0,
    "maxStakePerOracle": 0,
    "minStakeAmt": 0,
    "stakingEnabled": false,
    "totalRewardsToDistribute": 0,
  },
  "subscription": 0,
  "userAward": 0,
  "userStake": Object {},
}
`;

exports[`stake reducer oracleStake 1`] = `
Object {
  "addresses": Object {
    "operator": "0x0",
    "priceFeedList": Array [],
  },
  "allowance": 1000,
  "contractDUO": 0,
  "duo": 1000,
  "oracleStake": Object {},
  "stagingAdd": Object {},
  "states": Object {
    "canStake": false,
    "canUnstake": false,
    "lockMinTimeInSecond": 300,
    "maxStakePerOracle": 100000,
    "minStakeAmt": 500,
    "stakingEnabled": false,
    "totalAwardsToDistribute": 0,
    "totalRewardsToDistribute": 0,
  },
  "subscription": undefined,
  "userAward": 0,
  "userStake": Object {},
}
`;

exports[`stake reducer stakeAddress 1`] = `
Object {
  "addresses": Object {
    "operator": "0x0",
    "priceFeedList": Array [],
  },
  "allowance": 0,
  "contractDUO": 0,
  "duo": 0,
  "oracleStake": Object {},
  "stagingAdd": Object {},
  "states": Object {
    "canStake": false,
    "canUnstake": false,
    "lockMinTimeInSecond": 300,
    "maxStakePerOracle": 100000,
    "minStakeAmt": 500,
    "stakingEnabled": false,
    "totalAwardsToDistribute": 0,
    "totalRewardsToDistribute": 0,
  },
  "subscription": 0,
  "userAward": 0,
  "userStake": Object {},
}
`;

exports[`stake reducer stakeAllowance 1`] = `
Object {
  "addresses": Object {
    "operator": "0x0",
    "priceFeedList": Array [],
  },
  "allowance": 1000,
  "contractDUO": 0,
  "duo": 1000,
  "oracleStake": Object {},
  "stagingAdd": Object {},
  "states": Object {
    "canStake": false,
    "canUnstake": false,
    "lockMinTimeInSecond": 300,
    "maxStakePerOracle": 100000,
    "minStakeAmt": 500,
    "stakingEnabled": false,
    "totalAwardsToDistribute": 0,
    "totalRewardsToDistribute": 0,
  },
  "subscription": 0,
  "userAward": 0,
  "userStake": Object {},
}
`;

exports[`stake reducer stakeBalance 1`] = `
Object {
  "addresses": Object {
    "operator": "0x0",
    "priceFeedList": Array [],
  },
  "allowance": 0,
  "contractDUO": 0,
  "duo": 1000,
  "oracleStake": Object {},
  "stagingAdd": Object {},
  "states": Object {
    "canStake": false,
    "canUnstake": false,
    "lockMinTimeInSecond": 300,
    "maxStakePerOracle": 100000,
    "minStakeAmt": 500,
    "stakingEnabled": false,
    "totalAwardsToDistribute": 0,
    "totalRewardsToDistribute": 0,
  },
  "subscription": 0,
  "userAward": 0,
  "userStake": Object {},
}
`;

exports[`stake reducer stakeStates 1`] = `
Object {
  "addresses": Object {
    "operator": "0x0",
    "priceFeedList": Array [],
  },
  "allowance": 0,
  "contractDUO": 0,
  "duo": 0,
  "oracleStake": Object {},
  "stagingAdd": Object {},
  "states": Object {
    "canStake": false,
    "canUnstake": false,
    "lockMinTimeInSecond": 300,
    "maxStakePerOracle": 100000,
    "minStakeAmt": 500,
    "stakingEnabled": false,
    "totalAwardsToDistribute": 0,
    "totalRewardsToDistribute": 0,
  },
  "subscription": 0,
  "userAward": 0,
  "userStake": Object {},
}
`;

exports[`stake reducer stakeSub Off 1`] = `
Object {
  "addresses": Object {
    "operator": "0x0",
    "priceFeedList": Array [],
  },
  "allowance": 1000,
  "contractDUO": 0,
  "duo": 1000,
  "oracleStake": Object {},
  "stagingAdd": Object {},
  "states": Object {
    "canStake": false,
    "canUnstake": false,
    "lockMinTimeInSecond": 300,
    "maxStakePerOracle": 100000,
    "minStakeAmt": 500,
    "stakingEnabled": false,
    "totalAwardsToDistribute": 0,
    "totalRewardsToDistribute": 0,
  },
  "subscription": undefined,
  "userAward": 0,
  "userStake": Object {},
}
`;

exports[`stake reducer stakeSub On 1`] = `
Object {
  "addresses": Object {
    "operator": "0x0",
    "priceFeedList": Array [],
  },
  "allowance": 1000,
  "contractDUO": 0,
  "duo": 1000,
  "oracleStake": Object {},
  "stagingAdd": Object {},
  "states": Object {
    "canStake": false,
    "canUnstake": false,
    "lockMinTimeInSecond": 300,
    "maxStakePerOracle": 100000,
    "minStakeAmt": 500,
    "stakingEnabled": false,
    "totalAwardsToDistribute": 0,
    "totalRewardsToDistribute": 0,
  },
  "subscription": undefined,
  "userAward": 0,
  "userStake": Object {},
}
`;

exports[`stake reducer userAward 1`] = `
Object {
  "addresses": Object {
    "operator": "0x0",
    "priceFeedList": Array [],
  },
  "allowance": 1000,
  "contractDUO": 0,
  "duo": 1000,
  "oracleStake": Object {},
  "stagingAdd": Object {},
  "states": Object {
    "canStake": false,
    "canUnstake": false,
    "lockMinTimeInSecond": 300,
    "maxStakePerOracle": 100000,
    "minStakeAmt": 500,
    "stakingEnabled": false,
    "totalAwardsToDistribute": 0,
    "totalRewardsToDistribute": 0,
  },
  "subscription": undefined,
  "userAward": 0,
  "userStake": Object {},
}
`;

exports[`stake reducer userStake 1`] = `
Object {
  "addresses": Object {
    "operator": "0x0",
    "priceFeedList": Array [],
  },
  "allowance": 1000,
  "contractDUO": 0,
  "duo": 1000,
  "oracleStake": Object {},
  "stagingAdd": Object {},
  "states": Object {
    "canStake": false,
    "canUnstake": false,
    "lockMinTimeInSecond": 300,
    "maxStakePerOracle": 100000,
    "minStakeAmt": 500,
    "stakingEnabled": false,
    "totalAwardsToDistribute": 0,
    "totalRewardsToDistribute": 0,
  },
  "subscription": undefined,
  "userAward": 0,
  "userStake": Object {},
}
`;
